### PR TITLE
Gpio command controller

### PR DESCRIPTION
**Pull request by mcbed**
Controller to send commands to gpio interfaces allowing specific command interfaces for each gpio.

### CODE REVIEW

**Comment by mcbed** [comment]
Do you think there could be some interest in such a controller ? Currently I use it in galactic so is it interesting to adapt it for rolling ?

File: gpio_controllers/src/gpio_command_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    command_interfaces_config.names = command_interface_types_;
+
+    return command_interfaces_config;
+}
+
+controller_interface::InterfaceConfiguration
+GpioCommandController::state_interface_configuration() const
+{
+  return controller_interface::InterfaceConfiguration{
+    controller_interface::interface_configuration_type::NONE};
```
Comment (destogl) [inline gpio_controllers/src/gpio_command_controller.cpp:108]: Do we also want to expose `inputs` to the hardware as states (outputs) from the controller?

**Review by destogl — COMMENTED** [review]
Thank you for the great proposal!

Just few comments to discuss, making this even more generic.

File: gpio_controllers/src/gpio_command_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+CallbackReturn GpioCommandController::on_init()
+{
+  try
+  {
+    auto_declare<std::vector<std::string>>("gpios", std::vector<std::string>());
+    gpio_names_ = node_->get_parameter("gpios").as_string_array();
+    for(std::string &gpio : gpio_names_) 
+      auto_declare<std::vector<std::string>>("command_interfaces." + gpio, std::vector<std::string>());
```
[the remark this answers is not in the document]
Comment (mcbed) [reply gpio_controllers/src/gpio_command_controller.cpp]: Ok, this is done. There are still some fails but I don't see how to fix them :
`gpio_command_controller.cpp:128:  Missing space around colon in range-based for loop  [whitespace/forcolon] [2]`. Any idea ?

File: gpio_controllers/src/gpio_command_controller.cpp (hunk 1 above)
Comment (mcbed) [reply gpio_controllers/src/gpio_command_controller.cpp:111]: Very good point. For now I am using the `\dynamic_joint_states` topic to do it.

File: gpio_controllers/doc/userdoc.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                    - Gpio1
+                    - Gpio2
+                command_interfaces: 
+                    Gpio1: 
+                        - dig.1
+                        - dig.2
+                        - dig.3
+                    Gpio2: 
+                        - ana.1
+                        - ana.2
```
[the remark this answers is not in the document]
Comment (mcbed) [reply gpio_controllers/doc/userdoc.rst]: The controller is inspired from the multi interface forward controller PR [154](https://github.com/ros-controls/ros2_controllers/pull/154) so each GPIO that can have multiple ports can also have multiple specific interfaces. For example if Gpio1 is a module that has 3 digital outputs they can be claimed by the controller with the specified `command_interfaces`. Does this answer your first point ?

[the remark this answers is not in the document]
Comment (destogl) [reply gpio_controllers/doc/userdoc.rst]: Yes, something like that would be possible. What we could additionally do, add `size` parameter so one ve to write only the base name and size and not the same thing seventeen times. What do you think? Something like this:

```
gpios: 
    - gpio
    - speed_scaling
    - resend_robot_program
    - payload
command_interfaces:
    gpio:
      base_name: standard_digital_output_cmd
      size:  17
    speed_scaling:
        - target_speed_fraction_cmd
        - target_speed_fraction_async_success
    resend_robot_program: 
```

Should we limit this  `base_name` and `size` to the parameter `gpio`? I am uncertain if I find this very clean. What to do with digital inputs and analog IOs? Do we need more complex structure?

File: gpio_controllers/src/gpio_command_controller.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (destogl) [reply gpio_controllers/src/gpio_command_controller.cpp]: This should mean something like `for (auto const & name:names)` and it should be `for (auto const & name : names)`

File: gpio_controllers/src/gpio_command_controller.cpp (hunk 1 above)
Comment (destogl) [reply gpio_controllers/src/gpio_command_controller.cpp:111]: To be honest, I am not sure, what would you prefer? @livanov93 and @fmauch, do you have an opinion about this?

File: gpio_controllers/doc/userdoc.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcbed) [reply gpio_controllers/doc/userdoc.rst]: I agree that repeating 18 times the entry is repetitive but is it common in defining gpios ? In my usage I rather name these interfaces explicitly to now what is their purpose.

File: gpio_controllers/src/gpio_command_controller.cpp (hunk 1 above)
Comment (fmauch) [reply gpio_controllers/src/gpio_command_controller.cpp:111]: I am not deep enough (yet) to give a proper answer to this, I'm afraid.

However, having the state of IOs in general sounds like something that should be there. Not only things that are only inputs to the robot, but also the state of outputs (as they might be set by a different place (PLC, Teach pendant, hardware buttons, etc.))

File: gpio_controllers/include/gpio_controllers/gpio_command_controller.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+#include "rclcpp/subscription.hpp"
+#include "rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp"
+#include "rclcpp_lifecycle/state.hpp"
+#include "realtime_tools/realtime_buffer.h"
+#include "std_msgs/msg/float64_multi_array.hpp"
+
+namespace gpio_controllers
+{
+using CmdType = std_msgs::msg::Float64MultiArray;
+using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
```
[the remark this answers is not in the document]
Comment (mcbed) [reply gpio_controllers/include/gpio_controllers/gpio_command_controller.hpp:34]: I have a build issue when doing so: 
`error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type` 
This build issue is also present in the CI build for other controllers. Do I miss something ?

File: gpio_controllers/src/gpio_command_controller.cpp (hunk 1 above)
Comment (mcbed) [reply gpio_controllers/src/gpio_command_controller.cpp]: OK, I added the feature of publishing the state of commanded gpios as an `control_msgs::msg::InterfaceValue`.

**Comment by boilerbots** [comment]
I am building a system with Advantech EtherCAT modules and I want to use this controller or something similar. I think this gets a little confusing with comand_interfaces and state_interfaces.

What about dividing up Digital vs Analog interfaces instead? This would allow digital interfaces to have a single bit represent their state and floats for analog ports. The digital interfaces could define their direction and enforce that or just always be treated as bidirectional and depending on the hardware behind them you either read the value of the port or the internal state of the register.

The same thing would work for Analog ports.

**Comment by mcbed** [comment]
> I am building a system with Advantech EtherCAT modules and I want to use this controller or something similar. I think this gets a little confusing with comand_interfaces and state_interfaces.
> 
> What about dividing up Digital vs Analog interfaces instead? This would allow digital interfaces to have a single bit represent their state and floats for analog ports. The digital interfaces could define their direction and enforce that or just always be treated as bidirectional and depending on the hardware behind them you either read the value of the port or the internal state of the register.
> 
> The same thing would work for Analog ports.

Hi, I think that splitting Digital and Analog IO would reduce the generality of the controller but I am open to discussion.
Also, in its current version, the controller is intended to only provide commands to the interfaces tagged in the `command_interface` parameter. Btw, in all cases, all other GPIO states are published by the `joint_state_broadcaster`.  The `state_interface` flag does not exist in this controller, it was just a  discussion argument, but the feedback of all `command_interfaces` is published in the `~/gpio_states` topic as `control_msgs::msg::DynamicJointState`, thus my previous question: do you think it is the most relevant message type, even if the type joint is confusing, or should we add a new `control_msgs` type like `DynamicGPIOState`?

**Comment by christophfroehlich** [comment]
> Hi @destogl @mcbed,
> 
> I notice that the PR is still open, so I assume that this feature is not the top priority. If you are short on time and there are tasks that need completion, I am willing to continue working on this.

Let's see what the CI is telling us about compatibility now. 
We had some API breaking changes, at least they have to be addressed.
From what I see now, we also should add the generate_parameter_library for the parameter handling. @Wiktor-99 Do you want to take this over? Just start from this branch and open a new PR. If @mcbed is still around, you could also target your PR against his branch.

**Comment by Wiktor-99** [comment]
> > Hi @destogl @mcbed,
> > I notice that the PR is still open, so I assume that this feature is not the top priority. If you are short on time and there are tasks that need completion, I am willing to continue working on this.
> 
> Let's see what the CI is telling us about compatibility now. We had some API breaking changes, at least they have to be addressed. From what I see now, we also should add the generate_parameter_library for the parameter handling. @Wiktor-99 Do you want to take this over? Just start from this branch and open a new PR. If @mcbed is still around, you could also target your PR against his branch.



> > Hi @destogl @mcbed,
> > I notice that the PR is still open, so I assume that this feature is not the top priority. If you are short on time and there are tasks that need completion, I am willing to continue working on this.
> 
> Let's see what the CI is telling us about compatibility now. We had some API breaking changes, at least they have to be addressed. From what I see now, we also should add the generate_parameter_library for the parameter handling. @Wiktor-99 Do you want to take this over? Just start from this branch and open a new PR. If @mcbed is still around, you could also target your PR against his branch.

Yes, I would like to take this over if @mcbed is okay with it.

**Comment by christophfroehlich** [comment]
Superseded by #1251

**Pull request closed without merging**